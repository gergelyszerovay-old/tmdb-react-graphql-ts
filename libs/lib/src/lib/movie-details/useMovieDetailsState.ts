import {useQuery, useReactiveVar} from "@apollo/client";
import {QUERY_SELECTED_MOVIE} from "../state-gql/gql";
import {fetchStateVar, selectedMovieIdVar, similarMoviesListVar} from "../state-gql/cache";

export interface MovieDetailsState {
  onListSimilarMovies: () => void;
  selectedMovie: any;
  selectedMovieError: string;
}

const useMovieDetailsState = (testing = false): MovieDetailsState => {
  const fetchState = useReactiveVar(fetchStateVar);
  const selectedMovieId = useReactiveVar(selectedMovieIdVar);

  const {data: selectedMovie} = useQuery(QUERY_SELECTED_MOVIE, {
    skip: selectedMovieId === 0,
    variables: {id: selectedMovieId},
    fetchPolicy: testing ? "cache-first" : "cache-only" // https://github.com/apollographql/react-apollo/issues/779
    // fetchPolicy: "cache-only"
  });

  return {
    onListSimilarMovies: () => {
      similarMoviesListVar(selectedMovie.movies.movie.similar.edges);
    },
    selectedMovie: selectedMovie?.movies.movie,
    selectedMovieError: fetchState.useLazyQueryMovieDetails?.errorMessage || ''
  }
}

export default useMovieDetailsState;
