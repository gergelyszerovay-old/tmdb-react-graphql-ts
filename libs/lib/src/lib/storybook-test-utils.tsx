import React from 'react';
import {StylesProvider} from '@material-ui/styles';

// https://github.com/mui-org/material-ui/issues/9492
export const WithStableMuiClassnames = (c) => {
  const generateClassName = () => {
    let counter = 0
    return (rule, styleSheet) => `${styleSheet.options.classNamePrefix}-${rule.key}-${counter++}`
  }

  return (
    <StylesProvider generateClassName={generateClassName()}>
      {c}
    </StylesProvider>
  )
}
