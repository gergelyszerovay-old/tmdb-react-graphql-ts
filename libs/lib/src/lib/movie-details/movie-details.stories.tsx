import React from 'react';
import {MovieDetails} from './movie-details';

// @ts-ignore
import movie from './fixtures/movie.json';

import {MovieDetailsState} from "./useMovieDetailsState";
import {action} from "@storybook/addon-actions";

export default {
  component: MovieDetails,
  title: 'MovieDetails',
};

export const primary = () => {
  const state = (): MovieDetailsState => {
    return {
      onListSimilarMovies: action('onListSimilarMovies'),
      selectedMovieError: '',
      selectedMovie: movie
    }
  };
  return <MovieDetails hook={state}/>;
};

export const error = () => {
  const state = (): MovieDetailsState => {
    return {
      onListSimilarMovies: action('onListSimilarMovies'),
      selectedMovieError: 'Error',
      selectedMovie: undefined
    }
  };
  return <MovieDetails hook={state}/>;
};

export const empty = () => {
  const state = (): MovieDetailsState => {
    return {
      onListSimilarMovies: action('onListSimilarMovies'),
      selectedMovieError: '',
      selectedMovie: undefined
    }
  };
  return <MovieDetails hook={state}/>;
};

export const noWikipediaPageId = () => {
  const state = (): MovieDetailsState => {
    let m = {...movie};
    delete m.wikipediaPageId;
    return {
      onListSimilarMovies: action('onListSimilarMovies'),
      selectedMovieError: '',
      selectedMovie: m
    }
  };
  return <MovieDetails hook={state}/>;
};

export const noIMDBId = () => {
  const state = (): MovieDetailsState => {
    let m = {...movie};
    delete m.externalIds.imdb;
    return {
      onListSimilarMovies: action('onListSimilarMovies'),
      selectedMovieError: '',
      selectedMovie: m
    }
  };
  return <MovieDetails hook={state}/>;
};

export const noExternalIds = () => {
  const state = (): MovieDetailsState => {
    let m = {...movie};
    delete m.externalIds;
    return {
      onListSimilarMovies: action('onListSimilarMovies'),
      selectedMovieError: '',
      selectedMovie: m
    }
  };
  return <MovieDetails hook={state}/>;
};

export const noSimilarMovies = () => {
  const state = (): MovieDetailsState => {
    let m = {...movie};
    delete m.similar;
    return {
      onListSimilarMovies: action('onListSimilarMovies'),
      selectedMovieError: '',
      selectedMovie: m
    }
  };
  return <MovieDetails hook={state}/>;
};

export const noWikipediaExtract = () => {
  const state = (): MovieDetailsState => {
    let m = {...movie, wikipediaExtract: ''};
    return {
      onListSimilarMovies: action('onListSimilarMovies'),
      selectedMovieError: '',
      selectedMovie: m
    }
  };
  return <MovieDetails hook={state}/>;
};

export const noWikipediaExtractThreeDots = () => {
  const state = (): MovieDetailsState => {
    let m = {...movie, wikipediaExtract: '...'};
    return {
      onListSimilarMovies: action('onListSimilarMovies'),
      selectedMovieError: '',
      selectedMovie: m
    }
  };
  return <MovieDetails hook={state}/>;
};
