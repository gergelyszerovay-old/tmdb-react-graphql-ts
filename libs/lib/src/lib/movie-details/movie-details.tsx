import React, {FC} from 'react';

import './movie-details.scss';
import Button from "@material-ui/core/Button";
import useMovieDetailsState, {MovieDetailsState} from "./useMovieDetailsState";
import MuiAlert from "@material-ui/lab/Alert";
import {makeStyles} from "@material-ui/core/styles";

const useStyles = makeStyles((theme) => ({
  error: {
    margin: '10px'
  }
}));

/* eslint-disable-next-line */
export interface MovieDetailsProps {
  hook?: () => MovieDetailsState;
}

export const MovieDetails: FC<MovieDetailsProps> = ({hook = useMovieDetailsState}: MovieDetailsProps) => { // {movie, onListSimilarMovies}: MovieDetailsProps
  const classes = useStyles();
  const {selectedMovie, onListSimilarMovies, selectedMovieError} = hook();

  if (selectedMovieError) {
    return <MuiAlert elevation={6} variant="filled" severity="error" className={classes.error}>Can't load the movie
      details from the server!</MuiAlert>
  }

  if (!selectedMovie?.id) {
    return <React.Fragment/>
  }

  return (
    <div>
      <h1>{selectedMovie.title}</h1>
      {selectedMovie.wikipediaPageId ?
        <p><a href={"http://en.wikipedia.org/?curid=" + selectedMovie.wikipediaPageId} target="_blank"
              rel="noreferrer">Wikipedia</a></p> : <React.Fragment/>
      }
      {selectedMovie.externalIds?.imdb ?
        <p><a href={`https://www.imdb.com/title/${selectedMovie.externalIds.imdb}/`} target="_blank"
              rel="noreferrer">IMDB</a></p> : <React.Fragment/>
      }
      <div>
        {selectedMovie.similar?.edges ?
          <Button
            variant="contained"
            color="secondary"
            onClick={() => onListSimilarMovies()}
            data-testid="similar-movies-button"
          >
            List similar movies
          </Button> : <React.Fragment/>
        }
      </div>
      {selectedMovie.wikipediaExtract !== '' && selectedMovie.wikipediaExtract !== '...' ?
        <p>{selectedMovie.wikipediaExtract}</p> : <React.Fragment/>
      }
    </div>
  );
};

export default MovieDetails;
