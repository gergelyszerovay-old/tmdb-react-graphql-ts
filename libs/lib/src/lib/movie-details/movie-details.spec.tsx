import React from 'react';
import {fireEvent, getByTestId, render} from '@testing-library/react';

import MovieDetails from './movie-details';

import * as Stories from './movie-details.stories';
import {WithStableMuiClassnames} from "../storybook-test-utils";
import {MovieDetailsState} from "./useMovieDetailsState";

// @ts-ignore
import movie from './fixtures/movie.json';

describe('MovieDetails', () => {
  describe('On similar movies list click', () => {
    let container = null;
    let onListSimilarMoviesMock = null;

    beforeEach(() => {
      onListSimilarMoviesMock = jest.fn();
      const useMockState = (): MovieDetailsState => {
        return {
          onListSimilarMovies: onListSimilarMoviesMock,
          selectedMovie: movie,
          selectedMovieError: ''
        };
      };

      container = render(<MovieDetails hook={useMockState}/>).container;
    })

    it('Button should work', () => {
      const buttonEl = getByTestId(container, 'similar-movies-button');
      fireEvent.click(buttonEl);
      expect(onListSimilarMoviesMock).toHaveBeenCalled();
    });
  });


  describe('Storybook stories snapshot tests', () => {
    for (const [key, value] of Object.entries(Stories)) {
      if (value instanceof Function) {
        it(`Story ${key}`, () => {
          const {container} = render(WithStableMuiClassnames(Stories[key]()));
          expect(container.firstChild).toMatchSnapshot()
        });
      }
    }
  });

});
