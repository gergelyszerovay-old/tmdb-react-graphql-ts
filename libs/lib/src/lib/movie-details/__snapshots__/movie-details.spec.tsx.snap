// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MovieDetails Storybook stories snapshot tests Story empty 1`] = `null`;

exports[`MovieDetails Storybook stories snapshot tests Story error 1`] = `
<div
  class="MuiPaper-root-17 MuiAlert-root-1 MuiAlert-filledError-13 makeStyles-error-0 MuiPaper-elevation6-26"
  role="alert"
>
  <div
    class="MuiAlert-icon-14"
  >
    <svg
      aria-hidden="true"
      class="MuiSvgIcon-root-45 MuiSvgIcon-fontSizeInherit-51"
      focusable="false"
      viewBox="0 0 24 24"
    >
      <path
        d="M11 15h2v2h-2zm0-8h2v6h-2zm.99-5C6.47 2 2 6.48 2 12s4.47 10 9.99 10C17.52 22 22 17.52 22 12S17.52 2 11.99 2zM12 20c-4.42 0-8-3.58-8-8s3.58-8 8-8 8 3.58 8 8-3.58 8-8 8z"
      />
    </svg>
  </div>
  <div
    class="MuiAlert-message-15"
  >
    Can't load the movie details from the server!
  </div>
</div>
`;

exports[`MovieDetails Storybook stories snapshot tests Story noExternalIds 1`] = `
<div>
  <h1>
    Alien: Covenant
  </h1>
  <p>
    <a
      href="http://en.wikipedia.org/?curid=43219389"
      rel="noreferrer"
      target="_blank"
    >
      Wikipedia
    </a>
  </p>
  <div>
    <button
      class="MuiButtonBase-root-30 MuiButton-root-1 MuiButton-contained-9 MuiButton-containedSecondary-11"
      data-testid="similar-movies-button"
      tabindex="0"
      type="button"
    >
      <span
        class="MuiButton-label-2"
      >
        List similar movies
      </span>
      <span
        class="MuiTouchRipple-root-33"
      />
    </button>
  </div>
  <p>
    Alien: Covenant is a 2017 science fiction horror film directed and produced by Ridley Scott and written by John Logan and Dante Harper, from a story by Michael Green and Jack Paglen. A joint American and British production, the film is a sequel to Prometheus (2012) and is the second installment in the Alien prequel series and the sixth installment in the Alien franchise, as well as the third directed by Scott. The film features returning star Michael Fassbender and Katherine Waterston, with Billy Crudup, Danny McBride, and Demi치n Bichir in supporting roles. It follows the crew of a colony ship that lands on an uncharted planet and makes a terrifying discovery.
In 2012, prior to the release of Prometheus, Ridley Scott discussed the prospects of a sequel and new trilogy, and the film was confirmed that August. Principal photography began on April 4, 2016, at Milford Sound in Fiordland National Park, New Zealand, and wrapped on July 19, 2016. Effects houses Odd Studios and CreatureNFX provided...
  </p>
</div>
`;

exports[`MovieDetails Storybook stories snapshot tests Story noIMDBId 1`] = `
<div>
  <h1>
    Alien: Covenant
  </h1>
  <p>
    <a
      href="http://en.wikipedia.org/?curid=43219389"
      rel="noreferrer"
      target="_blank"
    >
      Wikipedia
    </a>
  </p>
  <div>
    <button
      class="MuiButtonBase-root-30 MuiButton-root-1 MuiButton-contained-9 MuiButton-containedSecondary-11"
      data-testid="similar-movies-button"
      tabindex="0"
      type="button"
    >
      <span
        class="MuiButton-label-2"
      >
        List similar movies
      </span>
      <span
        class="MuiTouchRipple-root-33"
      />
    </button>
  </div>
  <p>
    Alien: Covenant is a 2017 science fiction horror film directed and produced by Ridley Scott and written by John Logan and Dante Harper, from a story by Michael Green and Jack Paglen. A joint American and British production, the film is a sequel to Prometheus (2012) and is the second installment in the Alien prequel series and the sixth installment in the Alien franchise, as well as the third directed by Scott. The film features returning star Michael Fassbender and Katherine Waterston, with Billy Crudup, Danny McBride, and Demi치n Bichir in supporting roles. It follows the crew of a colony ship that lands on an uncharted planet and makes a terrifying discovery.
In 2012, prior to the release of Prometheus, Ridley Scott discussed the prospects of a sequel and new trilogy, and the film was confirmed that August. Principal photography began on April 4, 2016, at Milford Sound in Fiordland National Park, New Zealand, and wrapped on July 19, 2016. Effects houses Odd Studios and CreatureNFX provided...
  </p>
</div>
`;

exports[`MovieDetails Storybook stories snapshot tests Story noSimilarMovies 1`] = `
<div>
  <h1>
    Alien: Covenant
  </h1>
  <p>
    <a
      href="http://en.wikipedia.org/?curid=43219389"
      rel="noreferrer"
      target="_blank"
    >
      Wikipedia
    </a>
  </p>
  <div />
  <p>
    Alien: Covenant is a 2017 science fiction horror film directed and produced by Ridley Scott and written by John Logan and Dante Harper, from a story by Michael Green and Jack Paglen. A joint American and British production, the film is a sequel to Prometheus (2012) and is the second installment in the Alien prequel series and the sixth installment in the Alien franchise, as well as the third directed by Scott. The film features returning star Michael Fassbender and Katherine Waterston, with Billy Crudup, Danny McBride, and Demi치n Bichir in supporting roles. It follows the crew of a colony ship that lands on an uncharted planet and makes a terrifying discovery.
In 2012, prior to the release of Prometheus, Ridley Scott discussed the prospects of a sequel and new trilogy, and the film was confirmed that August. Principal photography began on April 4, 2016, at Milford Sound in Fiordland National Park, New Zealand, and wrapped on July 19, 2016. Effects houses Odd Studios and CreatureNFX provided...
  </p>
</div>
`;

exports[`MovieDetails Storybook stories snapshot tests Story noWikipediaExtract 1`] = `
<div>
  <h1>
    Alien: Covenant
  </h1>
  <p>
    <a
      href="http://en.wikipedia.org/?curid=43219389"
      rel="noreferrer"
      target="_blank"
    >
      Wikipedia
    </a>
  </p>
  <div>
    <button
      class="MuiButtonBase-root-30 MuiButton-root-1 MuiButton-contained-9 MuiButton-containedSecondary-11"
      data-testid="similar-movies-button"
      tabindex="0"
      type="button"
    >
      <span
        class="MuiButton-label-2"
      >
        List similar movies
      </span>
      <span
        class="MuiTouchRipple-root-33"
      />
    </button>
  </div>
</div>
`;

exports[`MovieDetails Storybook stories snapshot tests Story noWikipediaExtractThreeDots 1`] = `
<div>
  <h1>
    Alien: Covenant
  </h1>
  <p>
    <a
      href="http://en.wikipedia.org/?curid=43219389"
      rel="noreferrer"
      target="_blank"
    >
      Wikipedia
    </a>
  </p>
  <div>
    <button
      class="MuiButtonBase-root-30 MuiButton-root-1 MuiButton-contained-9 MuiButton-containedSecondary-11"
      data-testid="similar-movies-button"
      tabindex="0"
      type="button"
    >
      <span
        class="MuiButton-label-2"
      >
        List similar movies
      </span>
      <span
        class="MuiTouchRipple-root-33"
      />
    </button>
  </div>
</div>
`;

exports[`MovieDetails Storybook stories snapshot tests Story noWikipediaPageId 1`] = `
<div>
  <h1>
    Alien: Covenant
  </h1>
  <p>
    <a
      href="https://www.imdb.com/title/tt2316204/"
      rel="noreferrer"
      target="_blank"
    >
      IMDB
    </a>
  </p>
  <div>
    <button
      class="MuiButtonBase-root-30 MuiButton-root-1 MuiButton-contained-9 MuiButton-containedSecondary-11"
      data-testid="similar-movies-button"
      tabindex="0"
      type="button"
    >
      <span
        class="MuiButton-label-2"
      >
        List similar movies
      </span>
      <span
        class="MuiTouchRipple-root-33"
      />
    </button>
  </div>
  <p>
    Alien: Covenant is a 2017 science fiction horror film directed and produced by Ridley Scott and written by John Logan and Dante Harper, from a story by Michael Green and Jack Paglen. A joint American and British production, the film is a sequel to Prometheus (2012) and is the second installment in the Alien prequel series and the sixth installment in the Alien franchise, as well as the third directed by Scott. The film features returning star Michael Fassbender and Katherine Waterston, with Billy Crudup, Danny McBride, and Demi치n Bichir in supporting roles. It follows the crew of a colony ship that lands on an uncharted planet and makes a terrifying discovery.
In 2012, prior to the release of Prometheus, Ridley Scott discussed the prospects of a sequel and new trilogy, and the film was confirmed that August. Principal photography began on April 4, 2016, at Milford Sound in Fiordland National Park, New Zealand, and wrapped on July 19, 2016. Effects houses Odd Studios and CreatureNFX provided...
  </p>
</div>
`;

exports[`MovieDetails Storybook stories snapshot tests Story primary 1`] = `
<div>
  <h1>
    Alien: Covenant
  </h1>
  <p>
    <a
      href="http://en.wikipedia.org/?curid=43219389"
      rel="noreferrer"
      target="_blank"
    >
      Wikipedia
    </a>
  </p>
  <p>
    <a
      href="https://www.imdb.com/title/tt2316204/"
      rel="noreferrer"
      target="_blank"
    >
      IMDB
    </a>
  </p>
  <div>
    <button
      class="MuiButtonBase-root-30 MuiButton-root-1 MuiButton-contained-9 MuiButton-containedSecondary-11"
      data-testid="similar-movies-button"
      tabindex="0"
      type="button"
    >
      <span
        class="MuiButton-label-2"
      >
        List similar movies
      </span>
      <span
        class="MuiTouchRipple-root-33"
      />
    </button>
  </div>
  <p>
    Alien: Covenant is a 2017 science fiction horror film directed and produced by Ridley Scott and written by John Logan and Dante Harper, from a story by Michael Green and Jack Paglen. A joint American and British production, the film is a sequel to Prometheus (2012) and is the second installment in the Alien prequel series and the sixth installment in the Alien franchise, as well as the third directed by Scott. The film features returning star Michael Fassbender and Katherine Waterston, with Billy Crudup, Danny McBride, and Demi치n Bichir in supporting roles. It follows the crew of a colony ship that lands on an uncharted planet and makes a terrifying discovery.
In 2012, prior to the release of Prometheus, Ridley Scott discussed the prospects of a sequel and new trilogy, and the film was confirmed that August. Principal photography began on April 4, 2016, at Milford Sound in Fiordland National Park, New Zealand, and wrapped on July 19, 2016. Effects houses Odd Studios and CreatureNFX provided...
  </p>
</div>
`;
